File: src/frontend/contabil/lib/app/data/model/contabil_conta_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:contabil/app/infra/infra_imports.dart';
import 'package:contabil/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:contabil/app/data/domain/domain_imports.dart';

class ContabilContaModel {
	int? id;
	int? idPlanoConta;
	int? idPlanoContaRefSped;
	int? idContabilConta;
	String? classificacao;
	String? tipo;
	String? descricao;
	DateTime? dataInclusao;
	String? situacao;
	String? natureza;
	String? patrimonioResultado;
	String? livroCaixa;
	String? dfc;
	String? codigoEfd;
	String? ordem;
	String? codigoReduzido;
	PlanoContaModel? planoContaModel;
	PlanoContaRefSpedModel? planoContaRefSpedModel;

	ContabilContaModel({
		this.id,
		this.idPlanoConta,
		this.idPlanoContaRefSped,
		this.idContabilConta,
		this.classificacao,
		this.tipo,
		this.descricao,
		this.dataInclusao,
		this.situacao,
		this.natureza,
		this.patrimonioResultado,
		this.livroCaixa,
		this.dfc,
		this.codigoEfd,
		this.ordem,
		this.codigoReduzido,
		this.planoContaModel,
		this.planoContaRefSpedModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'id_contabil_conta',
		'classificacao',
		'tipo',
		'descricao',
		'data_inclusao',
		'situacao',
		'natureza',
		'patrimonio_resultado',
		'livro_caixa',
		'dfc',
		'codigo_efd',
		'ordem',
		'codigo_reduzido',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Id Contabil Conta',
		'Classificacao',
		'Tipo',
		'Descricao',
		'Data Inclusao',
		'Situacao',
		'Natureza',
		'Patrimonio Resultado',
		'Livro Caixa',
		'Dfc',
		'Codigo Efd',
		'Ordem',
		'Codigo Reduzido',
	];

	ContabilContaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idPlanoConta = jsonData['idPlanoConta'];
		idPlanoContaRefSped = jsonData['idPlanoContaRefSped'];
		idContabilConta = jsonData['idContabilConta'];
		classificacao = jsonData['classificacao'];
		tipo = ContabilContaDomain.getTipo(jsonData['tipo']);
		descricao = jsonData['descricao'];
		dataInclusao = jsonData['dataInclusao'] != null ? DateTime.tryParse(jsonData['dataInclusao']) : null;
		situacao = ContabilContaDomain.getSituacao(jsonData['situacao']);
		natureza = ContabilContaDomain.getNatureza(jsonData['natureza']);
		patrimonioResultado = ContabilContaDomain.getPatrimonioResultado(jsonData['patrimonioResultado']);
		livroCaixa = ContabilContaDomain.getLivroCaixa(jsonData['livroCaixa']);
		dfc = ContabilContaDomain.getDfc(jsonData['dfc']);
		codigoEfd = jsonData['codigoEfd'];
		ordem = jsonData['ordem'];
		codigoReduzido = jsonData['codigoReduzido'];
		planoContaModel = jsonData['planoContaModel'] == null ? PlanoContaModel() : PlanoContaModel.fromJson(jsonData['planoContaModel']);
		planoContaRefSpedModel = jsonData['planoContaRefSpedModel'] == null ? PlanoContaRefSpedModel() : PlanoContaRefSpedModel.fromJson(jsonData['planoContaRefSpedModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idPlanoConta'] = idPlanoConta != 0 ? idPlanoConta : null;
		jsonData['idPlanoContaRefSped'] = idPlanoContaRefSped != 0 ? idPlanoContaRefSped : null;
		jsonData['idContabilConta'] = idContabilConta;
		jsonData['classificacao'] = classificacao;
		jsonData['tipo'] = ContabilContaDomain.setTipo(tipo);
		jsonData['descricao'] = descricao;
		jsonData['dataInclusao'] = dataInclusao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataInclusao!) : null;
		jsonData['situacao'] = ContabilContaDomain.setSituacao(situacao);
		jsonData['natureza'] = ContabilContaDomain.setNatureza(natureza);
		jsonData['patrimonioResultado'] = ContabilContaDomain.setPatrimonioResultado(patrimonioResultado);
		jsonData['livroCaixa'] = ContabilContaDomain.setLivroCaixa(livroCaixa);
		jsonData['dfc'] = ContabilContaDomain.setDfc(dfc);
		jsonData['codigoEfd'] = codigoEfd;
		jsonData['ordem'] = ordem;
		jsonData['codigoReduzido'] = codigoReduzido;
		jsonData['planoContaModel'] = planoContaModel?.toJson;
		jsonData['planoContaRefSpedModel'] = planoContaRefSpedModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idPlanoConta = plutoRow.cells['idPlanoConta']?.value;
		idPlanoContaRefSped = plutoRow.cells['idPlanoContaRefSped']?.value;
		idContabilConta = plutoRow.cells['idContabilConta']?.value;
		classificacao = plutoRow.cells['classificacao']?.value;
		tipo = plutoRow.cells['tipo']?.value != '' ? plutoRow.cells['tipo']?.value : 'Sintética';
		descricao = plutoRow.cells['descricao']?.value;
		dataInclusao = Util.stringToDate(plutoRow.cells['dataInclusao']?.value);
		situacao = plutoRow.cells['situacao']?.value != '' ? plutoRow.cells['situacao']?.value : 'Ativa';
		natureza = plutoRow.cells['natureza']?.value != '' ? plutoRow.cells['natureza']?.value : 'Credora';
		patrimonioResultado = plutoRow.cells['patrimonioResultado']?.value != '' ? plutoRow.cells['patrimonioResultado']?.value : 'Patrimonio';
		livroCaixa = plutoRow.cells['livroCaixa']?.value != '' ? plutoRow.cells['livroCaixa']?.value : 'Sim';
		dfc = plutoRow.cells['dfc']?.value != '' ? plutoRow.cells['dfc']?.value : 'Não participa';
		codigoEfd = plutoRow.cells['codigoEfd']?.value;
		ordem = plutoRow.cells['ordem']?.value;
		codigoReduzido = plutoRow.cells['codigoReduzido']?.value;
		planoContaModel = PlanoContaModel();
		planoContaModel?.nome = plutoRow.cells['planoContaModel']?.value;
		planoContaRefSpedModel = PlanoContaRefSpedModel();
		planoContaRefSpedModel?.codCtaRef = plutoRow.cells['planoContaRefSpedModel']?.value;
	}	

	ContabilContaModel clone() {
		return ContabilContaModel(
			id: id,
			idPlanoConta: idPlanoConta,
			idPlanoContaRefSped: idPlanoContaRefSped,
			idContabilConta: idContabilConta,
			classificacao: classificacao,
			tipo: tipo,
			descricao: descricao,
			dataInclusao: dataInclusao,
			situacao: situacao,
			natureza: natureza,
			patrimonioResultado: patrimonioResultado,
			livroCaixa: livroCaixa,
			dfc: dfc,
			codigoEfd: codigoEfd,
			ordem: ordem,
			codigoReduzido: codigoReduzido,
		);			
	}

	
}

File: backend/contabil/src/model/contabil_conta_model.py
from src import db
from src.model.plano_conta_model import PlanoContaModel
from src.model.plano_conta_ref_sped_model import PlanoContaRefSpedModel


class ContabilContaModel(db.Model):
    __tablename__ = 'contabil_conta'

    id = db.Column(db.Integer, primary_key=True)
    id_contabil_conta = db.Column(db.Integer)
    classificacao = db.Column(db.String(30))
    tipo = db.Column(db.String(1))
    descricao = db.Column(db.String(100))
    data_inclusao = db.Column(db.DateTime)
    situacao = db.Column(db.String(1))
    natureza = db.Column(db.String(1))
    patrimonio_resultado = db.Column(db.String(1))
    livro_caixa = db.Column(db.String(1))
    dfc = db.Column(db.String(1))
    codigo_efd = db.Column(db.String(2))
    ordem = db.Column(db.String(20))
    codigo_reduzido = db.Column(db.String(10))
    id_plano_conta = db.Column(db.Integer, db.ForeignKey('plano_conta.id'))
    id_plano_conta_ref_sped = db.Column(db.Integer, db.ForeignKey('plano_conta_ref_sped.id'))

    plano_conta_model = db.relationship('PlanoContaModel', foreign_keys=[id_plano_conta])
    plano_conta_ref_sped_model = db.relationship('PlanoContaRefSpedModel', foreign_keys=[id_plano_conta_ref_sped])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_plano_conta = data.get('idPlanoConta')
        self.id_plano_conta_ref_sped = data.get('idPlanoContaRefSped')
        self.id_contabil_conta = data.get('idContabilConta')
        self.classificacao = data.get('classificacao')
        self.tipo = data.get('tipo')
        self.descricao = data.get('descricao')
        self.data_inclusao = data.get('dataInclusao')
        self.situacao = data.get('situacao')
        self.natureza = data.get('natureza')
        self.patrimonio_resultado = data.get('patrimonioResultado')
        self.livro_caixa = data.get('livroCaixa')
        self.dfc = data.get('dfc')
        self.codigo_efd = data.get('codigoEfd')
        self.ordem = data.get('ordem')
        self.codigo_reduzido = data.get('codigoReduzido')

    def serialize(self):
        return {
            'id': self.id,
            'idPlanoConta': self.id_plano_conta,
            'idPlanoContaRefSped': self.id_plano_conta_ref_sped,
            'idContabilConta': self.id_contabil_conta,
            'classificacao': self.classificacao,
            'tipo': self.tipo,
            'descricao': self.descricao,
            'dataInclusao': self.data_inclusao.isoformat(),
            'situacao': self.situacao,
            'natureza': self.natureza,
            'patrimonioResultado': self.patrimonio_resultado,
            'livroCaixa': self.livro_caixa,
            'dfc': self.dfc,
            'codigoEfd': self.codigo_efd,
            'ordem': self.ordem,
            'codigoReduzido': self.codigo_reduzido,
            'planoContaModel': self.plano_conta_model.serialize() if self.plano_conta_model else None,
            'planoContaRefSpedModel': self.plano_conta_ref_sped_model.serialize() if self.plano_conta_ref_sped_model else None,
        }